internal/auth/issuer/pam/pam_issuer.go

```diff
@@ -0,0 +1,951 @@
+	switch req.GrantType {
+	case pamapi.RefreshToken:
+		return s.handleRefreshTokenGrant(ctx, req)
+	case pamapi.AuthorizationCode:
+		return s.handleAuthorizationCodeGrant(ctx, req)
+	default:
+		return &pamapi.TokenResponse{Error: lo.ToPtr("unsupported_grant_type")}, nil
+	}
+}
+
+// handleRefreshTokenGrant handles the refresh_token grant type
+func (s *PAMOIDCProvider) handleRefreshTokenGrant(ctx context.Context, req *pamapi.TokenRequest) (*pamapi.TokenResponse, error) {
+	// Validate required fields for refresh token flow
+	if req.RefreshToken == nil || *req.RefreshToken == "" {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_request")}, nil
+	}
+
+	// Validate the refresh token and ensure it's actually a refresh token
+	identity, err := s.jwtGenerator.ValidateTokenWithType(*req.RefreshToken, "refresh_token")
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_grant")}, nil
+	}
+
+	// Get current user information from NSS to ensure user still exists
+	systemUser, err := s.pamAuthenticator.LookupUser(identity.GetUsername())
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("invalid_grant")}, nil
+	}
+
+	// Get current user groups for roles
+	groups, err := s.pamAuthenticator.GetUserGroups(systemUser)
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+	}
+
+	// Map groups to roles and extract organizations
+	roles := s.mapGroupsToRoles(groups)
+	organizations := s.extractOrganizations(groups)
+	tokenGenerationRequest := authn.TokenGenerationRequest{
+		Username:      identity.GetUsername(),
+		UID:           identity.GetUID(),
+		Organizations: organizations,
+		Roles:         roles,
+		Audience:      []string{s.config.ClientID}, // Set audience to client ID
+		Issuer:        s.config.Issuer,             // Set issuer
+	}
+	// Generate new access token with proper expiry (1 hour)
+	accessToken, err := s.jwtGenerator.GenerateTokenWithType(tokenGenerationRequest, time.Hour, "access_token")
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+	}
+
+	// Create token response
+	tokenResponse := &pamapi.TokenResponse{
+		AccessToken: lo.ToPtr(accessToken),
+		TokenType:   lo.ToPtr(pamapi.Bearer),
+		ExpiresIn:   lo.ToPtr(int(time.Hour.Seconds())),
+	}
+
+	// Always issue a new refresh token when using refresh_token grant
+	// (if we have a refresh token, it means offline_access was originally granted)
+	refreshToken, err := s.jwtGenerator.GenerateTokenWithType(tokenGenerationRequest, 7*24*time.Hour, "refresh_token")
+	if err != nil {
+		return &pamapi.TokenResponse{Error: lo.ToPtr("server_error")}, nil
+	}
+	tokenResponse.RefreshToken = lo.ToPtr(refreshToken)
+
+	return tokenResponse, nil
+}
+
+// handleAuthorizationCodeGrant handles the authorization_code grant type
+func (s *PAMOIDCProvider) handleAuthorizationCodeGrant(ctx context.Context, req *pamapi.TokenRequest) (*pamapi.TokenResponse, error) {
```

Comment:
is it ever possible req is nil? If we follow golangs error patterns my expectation is no.